File: 925-construct-binary-tree-from-preorder-and-postorder-traversal/construct-binary-tree-from-preorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructFromPrePost(self, a: List[int], b: List[int]) -> Optional[TreeNode]:
        n = len(a)
        idx = {val: i for i, val in enumerate(b)}  # Map values to postorder indices
        preIdx = 0  # Track current root index in preorder
        
        def cons(s, e):
            nonlocal preIdx
            if s > e:
                return None
            
            root = TreeNode(a[preIdx])
            preIdx += 1
            if s == e:
                return root  # Leaf node

            # Left subtree root is a[preIdx]
            leftSubtreeIdx = idx[a[preIdx]]  # Get its position in postorder
            
            # Recursively construct left and right subtrees
            root.left = cons(s, leftSubtreeIdx)
            root.right = cons(leftSubtreeIdx + 1, e - 1)  # Exclude root itself

            return root
        
        return cons(0, n - 1)

File: 925-construct-binary-tree-from-preorder-and-postorder-traversal/construct-binary-tree-from-preorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int n;
    
    public TreeNode constructFromPrePost(int[] preorder, int[] postorder) {
        n = preorder.length;
        return construct(0, n - 1, 0, n - 1, preorder, postorder);
    }

    private TreeNode construct(int l, int r, int lx, int rx, int[] pre, int[] pos) {
        if (l > r || lx > rx) return null;
        if (l == r) return new TreeNode(pre[l]);

        TreeNode head = new TreeNode(pre[l]);
        int rx_ = lx;
        int r_ = l + 1;

        while (pre[l + 1] != pos[rx_]) {
            rx_++;
            r_++;
        }

        head.left = construct(l + 1, r_, lx, rx_, pre, pos);
        if (r_ + 1 <= r) { // Ensure bounds are valid
            head.right = construct(r_ + 1, r, rx_ + 1, rx, pre, pos);
        }

        return head;
    }
}
